### PR TITLE
Add energon dataset support for Qwen2VL

### PR DESCRIPTION
**Pull request by romanbrickie**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Add energon dataset support for Qwen2VL.

**Collection**: [Note which collection this PR will affect]

# Changelog

- Add seperate module for energon dataset support
- Add 'energon' option for qwen2vl finetune script

# Usage

- torchrun --nproc_per_node=1 /opt/NeMo/scripts/vlm/qwen2vl_finetune.py    --num_nodes=1 --devices=1   --tp_size=1 --pp_size=1   --gbs=1 --mbs=1   --max_steps=10   --max_sequence_length=38400   --restore_path="/home/jepeng/projects/qwen2-vl/loss_convergence/qwen2vl2B_nemo_ckpt"   --projector_type="mlp2x_gelu"   --lr=0.00005   --data_type="energon"    --data_path="/home/jepeng/projects/qwen2-vl/rebase-nemo-qwen2vl/dev_utils/datasets/antique/short-complex-wds"   --image_folder=""   --wandb_project="nemo-qwen2vl-energon-loss-jepeng"   --name="_blending_energon_tp1pp1_gbs1_mbs1"   --log_dir="/workspace/logs"

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/collections/vlm/qwen2vl/data/convert_to_qwen2vl_wds.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+                conv = json.dumps(entry['conversations']).encode("utf-8")
+            elif 'messages' in entry:
+                conv = json.dumps(entry['messages']).encode("utf-8")
+            else:
+                conv = None
+            assert conv is not None, "No conversation texts"
+
+            sample = {
+                "__key__": entry.pop('id', str(idx)),
+                "jpgs": image_datas,
```
Comment (yaoyu-33) [inline nemo/collections/vlm/qwen2vl/data/convert_to_qwen2vl_wds.py:85]: rename to images_data and videos_data

File: nemo/collections/vlm/qwen2vl/data/energon.py [hunk 2]
```diff
@@ -0,0 +1,475 @@
...
+
+    cookers = [
+        Cooker(cook_chatml_sample),
+    ]
+
+    def __init__(
+        self,
+        tokenizer,
+        image_processor,
+        # data_paths: str | List[str],
```
Comment (yaoyu-33) [inline nemo/collections/vlm/qwen2vl/data/energon.py:158]: please clean all debugging comments

File: nemo/collections/vlm/qwen2vl/data/energon.py [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+        if video_token_indices is not None and video_thw_grids is not None:
+            assert len(video_token_indices) == len(
+                video_thw_grids
+            ), f"With {len(video_thw_grids)} videos in the sample, but {len(video_token_indices)} video placeholders!"
+        if image_thw_grids is not None and video_thw_grids is not None:
+            image_thw_grids, video_thw_grids = np.array(image_thw_grids, dtype=np.int64), np.array(
+                video_thw_grids, dtype=np.int64
+            )
+
+            target_length = (
```
Comment (yaoyu-33) [inline nemo/collections/vlm/qwen2vl/data/energon.py:293]: is this part from HF?

File: scripts/vlm/qwen2vl_finetune.py [hunk 4]
```diff
@@ -70,6 +73,7 @@ def main(args):
 
     max_sequence_length = args.max_sequence_length
     tokenizer = Qwen2Tokenizer.from_pretrained(hf_model_name)
+    image_processor = Qwen2VLImageProcessor()
     if args.data_type == "qwen2vl":
```
Comment (yaoyu-33) [inline scripts/vlm/qwen2vl_finetune.py:77]: can we rename it to `preloaded`?

File: scripts/vlm/qwen2vl_finetune.py [hunk 5]
```diff
@@ -70,6 +73,7 @@ def main(args):
 
     max_sequence_length = args.max_sequence_length
     tokenizer = Qwen2Tokenizer.from_pretrained(hf_model_name)
```
Comment (yaoyu-33) [inline scripts/vlm/qwen2vl_finetune.py:75]: This might break checkpoint saving? I think we need to wrap it with our AutoTokenizer `from nemo.collections.common.tokenizers import AutoTokenizer` and use `AutoTokenizer(hf_model_name)`

File: scripts/vlm/qwen2vl_finetune.py (hunk 4 above)
Comment (romanbrickie) [reply scripts/vlm/qwen2vl_finetune.py:77]: it is from Huggingface, better to keep as it is?

File: scripts/vlm/qwen2vl_finetune.py (hunk 5 above)
Comment (romanbrickie) [reply scripts/vlm/qwen2vl_finetune.py:75]: Tried to use AutoTokenizer, but error prompts. Let change to Auto after encode() function is added
image

Comment (romanbrickie) [reply scripts/vlm/qwen2vl_finetune.py:75]: and another issue
image

File: nemo/collections/vlm/qwen2vl/data/energon.py (hunk 3 above)
Comment (romanbrickie) [reply nemo/collections/vlm/qwen2vl/data/energon.py:285]: refer to pai-megatron-patch

File: nemo/collections/vlm/qwen2vl/data/energon.py (hunk 2 above)
Comment (romanbrickie) [reply nemo/collections/vlm/qwen2vl/data/energon.py]: done

File: nemo/collections/vlm/qwen2vl/data/convert_to_qwen2vl_wds.py (hunk 1 above)
Comment (romanbrickie) [reply nemo/collections/vlm/qwen2vl/data/convert_to_qwen2vl_wds.py]: done

File: nemo/collections/vlm/qwen2vl/data/convert_to_qwen2vl_wds.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import os
+import pickle
+from argparse import ArgumentParser
+
+import webdataset as wds
+from torchvision.io import encode_jpeg
```
[the remark this answers is not in the document]
Comment (romanbrickie) [reply nemo/collections/vlm/qwen2vl/data/convert_to_qwen2vl_wds.py]: resolved

File: nemo/collections/vlm/qwen2vl/data/convert_to_qwen2vl_wds.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+from tqdm import tqdm
+from webdataset.writer import add_handlers, default_handlers, imageencoder
```
[the remark this answers is not in the document]
Comment (romanbrickie) [reply nemo/collections/vlm/qwen2vl/data/convert_to_qwen2vl_wds.py]: resolved

File: nemo/collections/vlm/qwen2vl/data/convert_to_qwen2vl_wds.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+os.environ["FORCE_QWENVL_VIDEO_READER"] = 'torchvision'
+import numpy as np
+import torch
```
[the remark this answers is not in the document]
Comment (romanbrickie) [reply nemo/collections/vlm/qwen2vl/data/convert_to_qwen2vl_wds.py]: resolved

File: nemo/collections/vlm/qwen2vl/data/task_encoder.py [hunk 9]
```diff
@@ -0,0 +1,435 @@
...
+import dataclasses
+import pickle
+import re
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import Dict, List, Optional, Union
+
+import numpy as np
+import torch
+from megatron.energon import Batch, DefaultTaskEncoder, batch_list, batch_pad_stack
```
[the remark this answers is not in the document]
Comment (romanbrickie) [reply nemo/collections/vlm/qwen2vl/data/task_encoder.py]: resolved

File: nemo/collections/vlm/qwen2vl/data/task_encoder.py [hunk 10]
```diff
@@ -0,0 +1,435 @@
...
+from megatron.energon.flavors.base_dataset import Sample
+from megatron.energon.flavors.vqa import VQASample
```
[the remark this answers is not in the document]
Comment (romanbrickie) [reply nemo/collections/vlm/qwen2vl/data/task_encoder.py]: resolved

File: nemo/collections/vlm/qwen2vl/data/task_encoder.py [hunk 11]
```diff
@@ -0,0 +1,435 @@
...
+from megatron.energon.task_encoder.cooking import Cooker, basic_sample_keys
+from PIL import Image
+from torchvision.io import decode_jpeg
```
[the remark this answers is not in the document]
Comment (romanbrickie) [reply nemo/collections/vlm/qwen2vl/data/task_encoder.py]: resolved

File: nemo/collections/vlm/qwen2vl/data/task_encoder.py [hunk 12]
```diff
@@ -0,0 +1,435 @@
...
+from nemo.collections.vlm.qwen2vl.data.multimodal_tokens import (
+    HF_IMAGE_TOKEN_INDEX,
+    HF_VIDEO_TOKEN_INDEX,
+    IGNORE_INDEX,
+    IMAGE_TOKEN_INDEX,
+    PAD_TOKEN_INDEX,
+    SPECIAL_TOKEN_MAP,
+    VIDEO_TOKEN_INDEX,
+    VISION_END_TOKEN_INDEX,
+)
```
[the remark this answers is not in the document]
Comment (romanbrickie) [reply nemo/collections/vlm/qwen2vl/data/task_encoder.py]: resolved

**Review by yaoyu-33 — APPROVED** [review]
